base/abstractarray.jl

```diff
@@ -1095,55 +1095,76 @@ end
-    ndimsX = Int[isa(a,AbstractArray) ? ndims(a) : 0 for a in X]
-    ndimsC = max(maximum(ndimsX), maximum(catdims))
-    catsizes = zeros(Int,(nargs,length(catdims)))
-    dims2cat = zeros(Int,ndimsC)
-    for k = 1:length(catdims)
-        dims2cat[catdims[k]]=k
-    end
-
-    dimsC = Int[d <= ndimsX[1] ? size(X[1],d) : 1 for d=1:ndimsC]
-    for k = 1:length(catdims)
-        catsizes[1,k] = dimsC[catdims[k]]
-    end
-    for i = 2:nargs
-        for d = 1:ndimsC
-            currentdim = (d <= ndimsX[i] ? size(X[i],d) : 1)
-            if dims2cat[d] != 0
-                dimsC[d] += currentdim
-                catsizes[i,dims2cat[d]] = currentdim
-            elseif dimsC[d] != currentdim
-                throw(DimensionMismatch(string("mismatch in dimension ",d,
-                                               " (expected ",dimsC[d],
-                                               " got ",currentdim,")")))
+# helper functions
+cat_size(A) = (1,)
+cat_size(A::AbstractArray) = size(A)
+cat_size(A, d) = 1
+cat_size(A::AbstractArray, d) = size(A, d)
+
+cat_indices(A, d) = OneTo(1)
+cat_indices(A::AbstractArray, d) = indices(A, d)
+
+cat_similar(A, T, shape) = Array{T}(shape)
+cat_similar(A::AbstractArray, T, shape) = similar(A, T, shape)
+
+cat_shape(sifter, n, shape::Tuple) = shape
+@inline cat_shape(sifter, n, shape::Tuple, nshape::Tuple, shapes::Tuple...) =
+    cat_shape(sifter, n, _cshp(sifter, n, (), shape, nshape), shapes...)
+
+_cshp(::Tuple{}, n, out, ::Tuple{}, ::Tuple{}) = out
+_cshp(sifter, n, out, ::Tuple{}, ::Tuple{}) = (out..., map(b->(b ? n : 1), sifter)...)
+_cshp(::Tuple{}, n, out, shape, ::Tuple{}) = (out..., shape...)
+_cshp(::Tuple{}, n, out, ::Tuple{}, nshape) = (out..., nshape...)
+@inline _cshp(sifter, n, out, shape, ::Tuple{}) =
+    _cshp(tail(sifter), n, (out..., shape[1]), tail(shape), ())
+@inline _cshp(sifter, n, out, ::Tuple{}, nshape) =
+    _cshp(tail(sifter), n, (out..., nshape[1]), (), tail(nshape))
+@inline function _cshp(::Tuple{}, n, out, shape, nshape)
+    next = _cs(length(out) + 1, false, shape[1], nshape[1])
+    _cshp((), n, (out..., next), tail(shape), tail(nshape))
+end
+@inline function _cshp(sifter, n, out, shape, nshape)
+    next = _cs(length(out) + 1, sifter[1], shape[1], nshape[1])
+    _cshp(tail(sifter), n, (out..., next), tail(shape), tail(nshape))
+end
+
+_cs(d, concat, a, b) = concat ? (a + b) : (a == b ? a : throw(DimensionMismatch(string("mismatch in dimension ", d, " (expected ", a, " got ", b, ")"))))
+
+dims2sift{n}(::Type{Val{n}}) = ntuple(i -> (i == n), Val{n})
+dims2sift(dims) = ntuple(i -> (i in dims), maximum(dims))
+
+cat(dims, X...) = cat_t(dims, promote_eltype(X...), X...)
+
+function cat_t(dims, T::Type, X...)
+    sifter = dims2sift(dims)
+    shape = cat_shape(sifter, length(X), (), map(cat_size, X)...)
+    return _cat(T, shape, sifter, X...)
+end
+
+function _cat(T::Type, shape, sifter, X...)
+    N = length(shape)
+    A = cat_similar(X[1], T, shape)
```

Comment:
I'd like to handle that somehow, but I'd say that it would have to be addressed elsewhere. For one I have some ideas that would require the new sub-typing algorithm in place.